syntax = "proto3";
package metadata;

import "replica/replication.proto";

message NodeInfo {
  string node_id = 1;
  string host = 2;
  int32 port = 3;
}

message RegisterRequest {
  NodeInfo node = 1;
}

message HeartbeatRequest {
  string node_id = 1;
}

message ClusterState {
  repeated NodeInfo nodes = 1;
  replication.PartitionMap partition_map = 2;
}

service MetadataService {
  rpc RegisterNode(RegisterRequest) returns (ClusterState);
  rpc Heartbeat(HeartbeatRequest) returns (ClusterState);
  rpc GetClusterState(replication.Empty) returns (ClusterState);
  rpc UpdateClusterState(ClusterState) returns (replication.Empty);
}
